Wrappers/Python/ccpi/optimisation/functions/MixedL21Norm.py

```diff
@@ -19,105 +19,164 @@
-        ''' Evaluates L2,1Norm at point x
+        if self.b is not None and not isinstance(self.b, BlockDataContainer):
+            raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(self.b)))
             
-            :param: x is a BlockDataContainer
-                                
-        '''
+        
+    def __call__(self, x):
+        
+        r"""Returns the value of the MixedL21Norm function at x.                                            
+        """
         if not isinstance(x, BlockDataContainer):
             raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(x))) 
-                                         
-        tmp = x.get_item(0) * 0.
-        for el in x.containers:
-            tmp += el.power(2.)
-        return tmp.sqrt().sum()
+            
+#        tmp_cont = x.containers                        
+#        tmp = x.get_item(0) * 0.
+#        for el in tmp_cont:
+#            tmp += el.power(2.)
+#        tmp.add(self.epsilon**2, out = tmp)    
+#        return tmp.sqrt().sum()            
+         
+        y = x
+        if self.b is not None:
+            y = x - self.b    
+        return y.pnorm(p=2).sum()                                
+            
+#        tmp = x.get_item(0) * 0.
+#        for el in x.containers:
+#            tmp += el.power(2.)
+#        #tmp.add(self.epsilon, out = tmp)    
+#        return tmp.sqrt().sum()
 
                             
-    def gradient(self, x, out=None):
-        return ValueError('Not Differentiable')
-                            
     def convex_conjugate(self,x):
         
-        ''' This is the Indicator function of ||\cdot||_{2, \infty}
-            which is either 0 if ||x||_{2, \infty} or \infty        
-        '''
+        r"""Returns the value of the convex conjugate of the MixedL21Norm function at x.
         
-        return 0.0
+        This is the Indicator function of :math:`\mathbb{I}_{\{\|\cdot\|_{2,\infty}\leq1\}}(x^{*})`,
         
-    
+        i.e., 
+        
+        .. math:: \mathbb{I}_{\{\|\cdot\|_{2, \infty}\leq1\}}(x^{*}) 
+            = \begin{cases} 
+            0, \mbox{if } \|x\|_{2, \infty}\leq1\\
+            \infty, \mbox{otherwise}
+            \end{cases}
+        
+        where, 
+        
+        .. math:: \|x\|_{2,\infty} = \max\{ \|x\|_{2} \} = \max\{ \sqrt{ (x^{1})^{2} + (x^{2})^{2} + \dots}\}
+        
+        """
+        if not isinstance(x, BlockDataContainer):
+            raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(x))) 
+                
+#        tmp1 = x.get_item(0) * 0.
+#        for el in x.containers:
+#            tmp1 += el.power(2.)
+#        tmp1.add(self.epsilon**2, out = tmp1)
+#        tmp = tmp1.sqrt().as_array().max() - 1
+#                    
+#        if tmp<=1e-6:
+#            return 0
+#        else:
+#            return np.inf            
+                
+        tmp = (x.pnorm(2).as_array().max() - 1)
```

Comment:
implemented like this in order to avoid `infty` values when we evaluate primal-dual gap for the PDHG algorithm